replayer/frame_diff_serialization.cpp

```diff
@@ -10,90 +10,218 @@
 
 #include "frame.h"
+#include "flatbuffer_conversions.h"
+#include "streaming_flatbuffers.h"
 
 namespace torchcraft {
 namespace replayer {
 
-std::ostream& operator<<(std::ostream& out, const FrameDiff& o) {
-  out << o.pids.size();
-  for (auto& pid : o.pids)
-    out << " " << pid;
-  for (auto& player_unit : o.units) {
-    out << " " << player_unit.size();
-    for (auto& du : player_unit)
-      out << " " << du;
-  }
-  out << " " << o.creep_map.size();
-  for (auto pair : o.creep_map)
-    out << " " << pair.first << " " << pair.second;
-  out << " ";
-  writeTail(out, o.actions, o.resources, o.bullets);
-  out << " " << o.reward << " " << o.is_terminal;
+std::ostream& operator<<(std::ostream& out, const FrameDiff& frameDiff) {
+  flatbuffers::FlatBufferBuilder builder;
+  frameDiff.addToFlatBufferBuilder(builder);
+  writeFlatBufferToStream(out, builder);
   return out;
 }
 
-std::ostream& operator<<(
-    std::ostream& out,
-    const detail::UnitDiff& o) {
-  out << o.id << " " << o.velocityX << " " << o.velocityY;
-  out << " " << o.flags;
-  out << " " << o.var_ids.size();
-  for (size_t i = 0; i < o.var_ids.size(); i++)
-    out << " " << o.var_ids[i];
-  for (size_t i = 0; i < o.var_ids.size(); i++)
-    out << " " << o.var_diffs[i];
-  out << " " << o.order_size << " " << o.order_ids.size();
-  for (size_t i = 0; i < o.order_ids.size(); i++)
-    out << " " << o.order_ids[i];
-  for (size_t i = 0; i < o.order_ids.size(); i++)
-    out << " " << o.order_diffs[i];
-  return out;
+std::istream& operator>>(std::istream& in, FrameDiff& frameDiff) {
+  auto converter = [&frameDiff](const fbs::FrameDiff& fbsFrameDiff) {
+    frameDiff.readFromFlatBufferTable(fbsFrameDiff);
+  };
+  readFlatBufferTableFromStream<fbs::FrameDiff>(in, converter);
+  return in;
 }
 
-std::istream& operator>>(std::istream& in, FrameDiff& o) {
-  int32_t npids, n_creep_map_diff;
-  in >> npids;
-  o.pids.resize(npids);
-  o.units.resize(npids);
-  for (size_t i = 0; i < static_cast<size_t>(npids); i++)
-    in >> o.pids[i];
-  for (size_t i = 0, nunits = 0; i < static_cast<size_t>(npids); i++) {
-    in >> nunits;
-    o.units[i].resize(nunits);
-    for (size_t k = 0; k < nunits; k++)
-      in >> o.units[i][k];
-  }
-  in >> n_creep_map_diff;
-  for (size_t i = 0; i < static_cast<size_t>(n_creep_map_diff); i++) {
-    int32_t first, second;
-    in >> first >> second;
-    o.creep_map.insert(std::make_pair(first, second));
-  }
-  readTail(in, o.actions, o.resources, o.bullets);
-  in >> o.reward >> o.is_terminal;
-  return in;
+void FrameDiff::addToFlatBufferBuilder(flatbuffers::FlatBufferBuilder& builder) const {
+
+  auto buildFbsFrameDiffCreep = [](const std::pair<int32_t, int32_t> creepPair) {
+    return fbs::FrameDiffCreep(creepPair.first, creepPair.second);
+  };
+
+  auto buildFbsUnitDiffContainer = [&builder](const std::vector<detail::UnitDiff>& unitDiffs) {
+    auto buildFbsUnitDiff = [&builder](const detail::UnitDiff& unitDiff) {
```

Comment:
Why not like this?

```
return fbs::CreateUnitDiffDirect(
  builder,
  unitDiff.id,
  &unitDiff.var_ids,
  ...
);
```

Reply:
I find constructors with multiple arguments to be a maintenance antipattern, especially in languages without named parameters. One trap it presents is swapping the order of constructor arguments with the same type without swapping the parameters passed to it.

This approach takes the same number of lines of code and makes the correct mapping of value->parameter visually distinct.